pkg/gofr/datasource/oracle/oracle.go

```diff
@@ -0,0 +1,227 @@
+
+	return err
+}
+
+func (c *Client) Select(ctx context.Context, dest any, query string, args ...any) error {
+	tracedCtx, span := c.addTrace(ctx, "select", query)
+
+	err := c.conn.Select(tracedCtx, dest, query, args...)
+
+	defer c.sendOperationStats(time.Now(), "Select", query, "select", span, args...)
+
+	return err
+}
+
+func (c *Client) AsyncInsert(ctx context.Context, query string, wait bool, args ...any) error {
+	tracedCtx, span := c.addTrace(ctx, "async-insert", query)
+
+	err := c.conn.AsyncInsert(tracedCtx, query, wait, args...)
+
+	defer c.sendOperationStats(time.Now(), "AsyncInsert", query, "async-insert", span, args...)
+
+	return err
+}
+
+func (c *Client) sendOperationStats(start time.Time, methodType, query, method string, span trace.Span, args ...any) {
+	duration := time.Since(start).Microseconds()
+
+	c.logger.Debug(&Log{
+		Type:     methodType,
+		Query:    query,
+		Duration: duration,
+		Args:     args,
+	})
+
+	if span != nil {
+		defer span.End()
+		span.SetAttributes(attribute.Int64(fmt.Sprintf("oracle.%v.duration", method), duration))
+	}
+}
+
+type Health struct {
+	Status  string         `json:"status,omitempty"`
+	Details map[string]any `json:"details,omitempty"`
+}
+
+func (c *Client) HealthCheck(ctx context.Context) (any, error) {
+	h := Health{
+		Details: make(map[string]any),
+	}
+
+	h.Details["host"] = c.config.Host
+	h.Details["database"] = c.config.Service
+
+	err := c.conn.Ping(ctx)
+	if err != nil {
+		h.Status = "DOWN"
+		return &h, errStatusDown
+	}
+
+	h.Status = "UP"
+
+	return &h, nil
+}
+
+func (c *Client) addTrace(ctx context.Context, method, query string) (context.Context, trace.Span) {
+	if c.tracer != nil {
+		ctxWithTrace, span := c.tracer.Start(ctx, fmt.Sprintf("oracle-%v", method))
+
+		span.SetAttributes(
+			attribute.String("oracle.query", query),
+		)
+
+		return ctxWithTrace, span
+	}
+
+	return ctx, nil
+}
+
+type sqlConn struct{ db *sql.DB }
+
+func (s *sqlConn) Exec(ctx context.Context, query string, args ...any) error {
+	_, err := s.db.ExecContext(ctx, query, args...)
+	return err
+}
+
+func (s *sqlConn) Select(ctx context.Context, dest any, query string, args ...any) error {
+	rows, err := s.db.QueryContext(ctx, query, args...)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	columns, err := rows.Columns()
+	if err != nil {
+		return err
+	}
+
+	var results []map[string]any
+
+	for rows.Next() {
+		values := make([]any, len(columns))
+		valuePtrs := make([]any, len(columns))
+
+		for i := range columns {
+			valuePtrs[i] = &values[i]
+		}
+
+		if err := rows.Scan(valuePtrs...); err != nil {
+			return err
+		}
+
+		rowMap := make(map[string]any)
+		for i, col := range columns {
+			rowMap[col] = values[i]
+		}
+
+		results = append(results, rowMap)
+	}
+
+	if err := rows.Err(); err != nil {
+		return err
+	}
+
+	// Set the result to dest (must be *[]map[string]any).
+	p, ok := dest.(*[]map[string]any)
```

Comment:
Is it possible to typecast it to following ?
```suggestion
        type []map[string]any Destination
	p, ok := dest.(*Destination)
```

Reply:
I looked into updating the line from:

```go
p, ok := dest.(*[]map[string]any)
```

to

```go
type Destination []map[string]any
p, ok := dest.(*Destination)
```

However, i found out that Go treats `*Destination` and `*[]map[string]any` as two **distinct types**. Even though `Destination` is based on the same underlying slice type, the Go type system doesn’t allow a direct type assertion between a named type (`Destination`) and the unnamed slice type (`[]map[string]any`). This leads to a runtime type assertion failure. I have attached the screenshot of logs also when made the get request , when I applied your suggestion
<img width="672" height="86" alt="Screenshot 2025-07-19 000017" src="https://github.com/user-attachments/assets/01c25257-ff33-44af-9e4f-1186a1cf148a" />


Currently, most of the calling code passes `&[]map[string]any{}`. To safely use `*Destination`, we’d have to **refactor all upstream calls** (including tests) to use `*Destination{}` instead, which might introduce unnecessary overhead for now.